src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -0,0 +1,1449 @@
+
+#include <shader_ps.h>
+#include <shader_vs.h>
+
+#include "../../interactivity/win32/CustomWindowMessages.h"
+
+#define SHADER_SOURCE_DIRECTORY LR"(..\..\renderer\atlas\)"
+
+// #### NOTE ####
+// This file should only contain methods that are only accessed by the caller of Present() (the "Renderer" class).
+// Basically this file poses the "synchronization" point between the concurrently running
+// general IRenderEngine API (like the Invalidate*() methods) and the Present() method
+// and thus may access both _r and _api.
+
+#pragma warning(disable : 4100) // '...': unreferenced formal parameter
+// Disable a bunch of warnings which get in the way of writing performant code.
+#pragma warning(disable : 26429) // Symbol 'data' is never tested for nullness, it can be marked as not_null (f.23).
+#pragma warning(disable : 26446) // Prefer to use gsl::at() instead of unchecked subscript operator (bounds.4).
+#pragma warning(disable : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+#pragma warning(disable : 26482) // Only index into arrays using constant expressions (bounds.2).
+
+using namespace Microsoft::Console::Render;
+
+#pragma warning(push)
+#pragma warning(disable : 26447) // The function is declared 'noexcept' but calls function 'operator()()' which may throw exceptions (f.6).
+__declspec(noinline) static void showOOMWarning() noexcept
+{
+    [[maybe_unused]] static const auto once = []() {
+        std::thread t{ []() noexcept {
+            MessageBoxW(nullptr, L"This application is using a highly experimental text rendering engine and has run out of memory. Text rendering will start to behave irrationally and you should restart this process.", L"Out Of Memory", MB_ICONERROR | MB_OK);
+        } };
+        t.detach();
+        return false;
+    }();
+}
+#pragma warning(pop)
+
+struct TextAnalyzer final : IDWriteTextAnalysisSource, IDWriteTextAnalysisSink
+{
+    constexpr TextAnalyzer(const std::vector<wchar_t>& text, std::vector<AtlasEngine::TextAnalyzerResult>& results) noexcept :
+        _text{ text }, _results{ results }
+    {
+        Ensures(_text.size() <= UINT32_MAX);
+    }
+
+#ifndef NDEBUG
+private:
+    ULONG _refCount = 1;
+
+public:
+    ~TextAnalyzer()
+    {
+        assert(_refCount == 1);
+    }
+#endif
+
+    HRESULT __stdcall QueryInterface(const IID& riid, void** ppvObject) noexcept override
+    {
+        __assume(ppvObject != nullptr);
+
+        if (IsEqualGUID(riid, __uuidof(IDWriteTextAnalysisSource)) || IsEqualGUID(riid, __uuidof(IDWriteTextAnalysisSink)))
+        {
+            *ppvObject = this;
+            return S_OK;
+        }
+
+        *ppvObject = nullptr;
+        return E_NOINTERFACE;
+    }
+
+    ULONG __stdcall AddRef() noexcept override
+    {
+#ifdef NDEBUG
+        return 1;
+#else
+        return ++_refCount;
+#endif
+    }
+
+    ULONG __stdcall Release() noexcept override
+    {
+#ifdef NDEBUG
+        return 1;
+#else
+        return --_refCount;
+#endif

```

Comment:
Wait so you don't refcount when we're in release or what?

Reply:
No because that structure will be stack allocated only. There's no point to ref-count it then.

Reply:
Oh yeah ok that makes sense. Saving time. Can you leave that comment then because not refcounting a COM object is weird.